File: screener/screener.py
"""Core screener logic for MA Kuncup / MA Ketat pattern detection."""

from concurrent.futures import ThreadPoolExecutor

import pandas as pd

from screener.config import (
    RANGE_TICKS_THRESHOLD,
    VOL_PCT_THRESHOLD,
    MIN_VOLUME,
)
from screener.data import fetch_stock_data
from screener.indicators import calculate_all_indicators


OUTPUT_COLS = [
    "Ticker", "Close", "MA3", "MA5", "MA10", "MA20", "MA50", "MA100",
    "Range_Ticks", "Vol_Pct", "Volume", "MA_Tight", "Signal",
]


def is_ma_tight(range_ticks: float, vol_pct: float) -> bool:
    """Return True if the MA kuncup (tight) condition is satisfied."""
    return range_ticks < RANGE_TICKS_THRESHOLD and vol_pct < VOL_PCT_THRESHOLD


def is_signal(row: pd.Series) -> bool:
    """Return True if the final entry signal is triggered for a given row."""
    if pd.isna(row.get("MA_Tight")) or pd.isna(row.get("Volume")) or pd.isna(row.get("MA100")):
        return False
    return (
        bool(row["MA_Tight"])
        and row["Volume"] > MIN_VOLUME
        and row["MA100"] <= row["Close"]
    )


def _screen_single_stock_with_params(
    ticker: str,
    range_ticks_threshold: float,
    vol_pct_threshold: float,
    min_volume: int,
    min_rows: int,
) -> pd.Series | None:
    """Fetch/process one stock with caller-supplied thresholds."""
    df = fetch_stock_data(ticker)
    if df is None or df.empty or len(df) < min_rows:
        return None

    df = calculate_all_indicators(df)
    df["MA_Tight"] = (
        (df["Range_Ticks"] < range_ticks_threshold)
        & (df["Vol_Pct"] < vol_pct_threshold)
    )

    latest = df.iloc[-1].copy()
    latest["Signal"] = bool(
        pd.notna(latest.get("MA_Tight"))
        and pd.notna(latest.get("Volume"))
        and pd.notna(latest.get("MA100"))
        and bool(latest["MA_Tight"])
        and latest["Volume"] > min_volume
        and latest["MA100"] <= latest["Close"]
    )
    latest["Ticker"] = ticker
    return latest


def screen_single_stock(ticker: str) -> pd.Series | None:
    """Fetch and process a single stock with default config thresholds."""
    return _screen_single_stock_with_params(
        ticker=ticker,
        range_ticks_threshold=RANGE_TICKS_THRESHOLD,
        vol_pct_threshold=VOL_PCT_THRESHOLD,
        min_volume=MIN_VOLUME,
        min_rows=100,
    )


def run_screener(
    symbols: list,
    range_ticks_threshold: float = RANGE_TICKS_THRESHOLD,
    vol_pct_threshold: float = VOL_PCT_THRESHOLD,
    min_volume: int = MIN_VOLUME,
    min_rows: int = 100,
    max_workers: int = 4,
) -> pd.DataFrame:
    """Run the screener over all symbols and return a summary DataFrame."""
    if not symbols:
        return pd.DataFrame()

    def _worker(ticker: str) -> pd.Series | None:
        try:
            return _screen_single_stock_with_params(
                ticker=ticker,
                range_ticks_threshold=range_ticks_threshold,
                vol_pct_threshold=vol_pct_threshold,
                min_volume=min_volume,
                min_rows=min_rows,
            )
        except Exception as exc:  # noqa: BLE001
            print(f"[run_screener] Skip {ticker}: {exc}")
            return None

    worker_count = max(1, min(int(max_workers), len(symbols)))
    if worker_count == 1:
        results = (_worker(ticker) for ticker in symbols)
    else:
        with ThreadPoolExecutor(max_workers=worker_count) as executor:
            results = executor.map(_worker, symbols)

    records = [result for result in results if result is not None]

    if not records:
        return pd.DataFrame()

    df_out = pd.DataFrame(records)
    df_out = df_out[[c for c in OUTPUT_COLS if c in df_out.columns]].reset_index(drop=True)
    return df_out

File: vercel-screener-ma/lib/screener.ts
// Core screener logic

import { fetchStockData } from './data';
import { calculateAllIndicators, getLatestIndicators, StockDataWithIndicators } from './indicators';

export interface ScreenerResult {
  ticker: string;
  close: number;
  MA3: number;
  MA5: number;
  MA10: number;
  MA20: number;
  MA50: number;
  MA100: number;
  Range_Ticks: number;
  Vol_Pct: number;
  Volume: number;
  MA_Tight: boolean;
  Signal: boolean;
  error?: string;
}

/**
 * Screen a single stock
 * @param ticker - Stock ticker symbol
 * @param rangeTicks - Range ticks threshold
 * @param volPct - Volatility percentage threshold
 * @param minVolume - Minimum volume threshold
 * @returns Screener result or null
 */
export async function screenSingleStock(
  ticker: string,
  rangeTicks: number = 6,
  volPct: number = 3.8,
  minVolume: number = 1_000_000
): Promise<ScreenerResult | null> {
  try {
    const data = await fetchStockData(ticker);
    
    if (data.length < 100) {
      return {
        ticker,
        close: 0,
        MA3: 0,
        MA5: 0,
        MA10: 0,
        MA20: 0,
        MA50: 0,
        MA100: 0,
        Range_Ticks: 0,
        Vol_Pct: 0,
        Volume: 0,
        MA_Tight: false,
        Signal: false,
        error: 'Insufficient data',
      };
    }

    const dataWithIndicators = calculateAllIndicators(data, rangeTicks, volPct, minVolume);
    const latest = getLatestIndicators(dataWithIndicators);

    if (!latest) return null;

    return {
      ticker,
      close: latest.close,
      MA3: latest.MA3 || 0,
      MA5: latest.MA5 || 0,
      MA10: latest.MA10 || 0,
      MA20: latest.MA20 || 0,
      MA50: latest.MA50 || 0,
      MA100: latest.MA100 || 0,
      Range_Ticks: latest.Range_Ticks || 0,
      Vol_Pct: latest.Vol_Pct || 0,
      Volume: latest.volume,
      MA_Tight: latest.MA_Tight || false,
      Signal: latest.Signal || false,
    };
  } catch (error) {
    console.error(`Error screening ${ticker}:`, error);
    return null;
  }
}

/**
 * Run screener on multiple stocks
 * @param tickers - Array of ticker symbols
 * @param rangeTicks - Range ticks threshold
 * @param volPct - Volatility percentage threshold
 * @param minVolume - Minimum volume threshold
 * @param maxConcurrent - Maximum concurrent requests
 * @returns Array of screener results
 */
export async function runScreener(
  tickers: string[],
  rangeTicks: number = 6,
  volPct: number = 3.8,
  minVolume: number = 1_000_000,
  maxConcurrent: number = 10
): Promise<ScreenerResult[]> {
  const results: ScreenerResult[] = [];
  
  // Process in batches to avoid overwhelming the API
  for (let i = 0; i < tickers.length; i += maxConcurrent) {
    const batch = tickers.slice(i, i + maxConcurrent);
    const batchResults = await Promise.all(
      batch.map(ticker => screenSingleStock(ticker, rangeTicks, volPct, minVolume))
    );
    
    results.push(...batchResults.filter((r): r is ScreenerResult => r !== null));
  }

  return results;
}
